core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java

```diff
@@ -0,0 +1,114 @@
+package com.yahoo.imapnio.async.request;
+
+import javax.annotation.Nonnull;
+
+import com.yahoo.imapnio.async.data.MessageNumberSet;
+
+/**
+ * This class defines IMAP fetch command request from client. ABNF in RFC3501 is described as following:
+ *
+ * <pre>
+ * {@code
+ * fetch           = "FETCH" SP sequence-set SP ("ALL" / "FULL" / "FAST" /
+ *                   fetch-att / "(" fetch-att *(SP fetch-att) ")")
+ *
+ * fetch-att       = "ENVELOPE" / "FLAGS" / "INTERNALDATE" /
+ *                   "RFC822" [".HEADER" / ".SIZE" / ".TEXT"] /
+ *                   "BODY" ["STRUCTURE"] / "UID" /
+ *                   "BODY" section ["<" number "." nz-number ">"] /
+ *                   "BODY.PEEK" section ["<" number "." nz-number ">"]
+ * }
+ * </pre>
+ */
+public abstract class AbstractFetchCommand extends ImapRequestAdapter {
+
+    /** FETCH and space. */
+    private static final String FETCH_SPACE = "FETCH ";
+
+    /** FETCH and space. */
+    private static final String UID_FETCH_SPACE = "UID FETCH ";
+
+    /** Message numbers, either message sequence or UID. */
+    private String msgNumbers;
+
+    /** Fetch items. */
+    private String dataItems;
```

Comment:
Should we have separate classes one with message data items and another with macro?

Reply:
That would end up 2*2  classes (FetchMacroCommand,  FetchDataItemsCommand, UidFetchMacroCommand, UidFetchDataItemsCommand),  it is going to be a little too much to maintain.